### DISCUSSION TITLE
Is Float.precision the number of decimal or binary digits?

### DISCUSSION THREAD

**Discussion by eltoder** [sqlalchemy/sqlalchemy#11845] [Usage Questions]
The description of [`Float.precision`](https://docs.sqlalchemy.org/en/20/core/type_basics.html#sqlalchemy.types.Float.params.precision) makes it sound like it is the number of _decimal_ digits:
>Backends should attempt to ensure this precision indicates a number of digits for the generic Float datatype. For the Oracle backend, the Float.precision parameter is not accepted when rendering DDL, as Oracle does not support float precision specified as a number of decimal places. Instead, use the Oracle-specific FLOAT datatype and specify the FLOAT.binary_precision parameter.

However, [the text above it](https://docs.sqlalchemy.org/en/20/core/type_basics.html#sqlalchemy.types.Float) says:
>When a Float.precision is not provided in a Float type some backend may compile this type as an 8 bytes / 64 bit float datatype. To use a 4 bytes / 32 bit float datatype a precision <= 24 can usually be provided or the REAL type can be used.

Given that float32 has about 6 decimal digits of precision and float64 has about 15, but float32 does have 24 **bits** of mantissa, this appears to say that `precision` is measured in bits.

Moreover, for the PostgreSQL backend, Float is rendered like this in the DDL:
```python
return "FLOAT(%(precision)s)" % {"precision": type_.precision}
```
According to [the postgres documentation](https://www.postgresql.org/docs/current/datatype-numeric.html#DATATYPE-FLOAT), the value in parenthesis is the number of binary digits:
>Here, p specifies the minimum acceptable precision in binary digits.

So it sounds like the meaning of `Float.precision` is de facto the number of binary digits. Shall the documentation be reworded to make this clear?

**Comment by CaselIT** [comment]
Hi,

I think the description is consistent with the SQL standard and also what's mentioned in postgresql:
> Here, p specifies the minimum acceptable precision in binary digits. PostgreSQL accepts float(1) to float(24) as selecting the real type, while float(25) to float(53) select double precision. Values of p outside the allowed range draw an error. float with no precision specified is taken to mean double precision.

**Reply by eltoder** [reply]
The description says "precision indicates a number of digits". I think most people would interpret this as **decimal** digits, not as **binary** digits. If this is meant to be binary digits, it would be nice to clarify this explicitly. Notice how the PostgreSQL documentation explicitly says *binary* digits with an emphasis on *binary*. Also, if it is binary digits, it is unclear why the Oracle backend needs a separate `binary_precision`  parameter.

**Reply by CaselIT** [reply]
Oracle does things in it own way at the moment, see also https://github.com/sqlalchemy/sqlalchemy/issues/9704

It probably wouldn't hurt to specify that precision may refer to binary digits.

Personally I'm not sure how usual it is to specify the precision in the floats.

**Reply by eltoder** [reply]
Yes, Oracle stores floats in its own format with up to 38 decimal/126 binary digits of precision. It has separate "binary" float types for the more common IEEE types.

However, my point still stands -- Oracle's `FLOAT(p)` specifies a floating point value with p *binary* digits of precision, which is the same as postgres. So it is confusing that postgres uses the `precision` argument and Oracle uses `binary_precision` when they mean the same thing. Also, the explanation introducing `binary_precision` specifically refers to "decimal places", making it sound even more that `precision` is decimal.

 (Oracle's definition of `REAL` and `DOUBLE PRECISION` are different from Postgres': real has about 3x and double has over 2x more digits of precision. OTOH in `FLOAT(p)` both interpret `p` the same way.)

**Reply by CaselIT** [reply]
The difference is that postgresql seems to comply with the sql standard. oracle does not, hence the different parameter. 

See also https://learn.microsoft.com/en-us/sql/t-sql/data-types/float-and-real-transact-sql?view=sql-server-ver16 or https://dev.mysql.com/doc/refman/8.0/en/floating-point-types.html. Both behave like postgresql

**Reply by eltoder** [reply]
Could you clarify which behavior does not comply with which part of the SQL standard? Or what do you see as "the difference"? As I've said, they do the same thing.

Also, another quote from the documentation:
>To create a database agnostic Float that separately specifies binary precision for Oracle, use TypeEngine.with_variant() as follows:

```python
from sqlalchemy import Column
from sqlalchemy import Float
from sqlalchemy.dialects import oracle

Column(
    "float_data",
    Float(5).with_variant(oracle.FLOAT(binary_precision=16), "oracle")
)
```
This makes it look like Oracle's `binary_precision=16` is equivalent to `precision=5`, which implies that `precision` is decimal. If `precision` is also binary, the values should be the same.

**Reply by eltoder** [reply]
I think the confusion may come from the fact that Oracle defines REAL and DOUBLE PRECISION to have more precision than Postgres and many other dbs. So the values of precision in FLOAT needed to match REAL and DOUBLE are different. However, this is not because Oracle interprets precision differently, but because it defines REAL/DOUBLE differently. The precision is defined the same way.

**Reply by CaselIT** [reply]
> Could you clarify which behavior does not comply with which part of the SQL standard?

Most database have a float precision that does between 1 and 53. These are usually ignored and only used to decide if it will use float8 or float4. mysql, mssql, postgresql behave like this. this seems to be what the sql standard says.

Oracle float instead does something differente:
  > A subtype of the NUMBER data type having precision p. A FLOAT value is represented internally as NUMBER. The precision p can range from 1 to 126 binary digits. A FLOAT value requires from 1 to 22 bytes. 


> Or what do you see as "the difference"? As I've said, they do the same thing.

no they do not. please re-read what pg and oracle do. postgresql has only float4 and float8, it will not create a variable size float depending on the precision you specify. oracle does.

That documentation snippet illustrates how to customize things for oracle. I'm not sure if 5 and 16 have a particular meaning here, would need to look at the history to see if it was mentioned why the example was written like that.

I'm not sure what's the point of this discussion. Can you re-state what's wrong and what you would like to change?

**Reply by eltoder** [reply]
The SQL standard does not specify exact storage formats. This is what it says:
```
         21)FLOAT specifies the data type approximate numeric, with binary
            precision equal to or greater than the value of the specified
            <precision>. The maximum value of <precision> is implementation-
            defined. <precision> shall not be greater than this value.

         22)REAL specifies the data type approximate numeric, with implementation-
            defined precision.

         23)DOUBLE PRECISION specifies the data type approximate numeric,
            with implementation-defined precision that is greater than the
            implementation-defined precision of REAL.

         24)For the <approximate numeric type>s FLOAT, REAL, and DOUBLE
            PRECISION, the maximum and minimum values of the exponent are
            implementation-defined.
```
So the implementation is free to use variable-precision numbers with exactly `precision` bits (like Oracle does), or to round `precision` to a larger number (like postgres does). The maximum precision (126 for Oracle and 53 for postgres) is also implementation defined. So both behaviors comply with the standard. Also note that both Oracle and postgres interpret precision as binary, as required by the standard.

>I'm not sure if 5 and 16 have a particular meaning here, would need to look at the history to see if it was mentioned why the example was written like that.

The documentation for `binary_precision` says:
>This may be approximated to the number of decimal characters using the formula “decimal precision = 0.30103 * binary precision”.

5 ≈ 0.30103 * 16.

I am suggesting the following:

1. In the documentation for `Float.precision` clarify that it is the number of *binary* digits.
2. Use `precision` in the Oracle backend, because it means the same thing. Remove references to "decimal places" and the confusing example.
3. Ideally, rename `binary_precision` to just `precision` in `oracle.FLOAT`, though this requires a deprecation warning and may not be worth the effort.
4. In the paragraph in Float's documentation about selecting 4-byte floats, consider mentioning that Oracle floats are different and have different precision.

You raise a good point in #9704 that it may make sense to map Float/Double to Oracle's "binary" float/double, since this is more similar to other databases. People who want Oracle's actual `FLOAT` can use the UPPERCASE `FLOAT` type or `oracle.FLOAT`.

**Reply by CaselIT** [reply]
I'm not sure that the two mean the same thing with binary precision, since if we assume the oracle implementation is linear in the storage used, then 53 means 8bytes in most db, but oracle will uses 10 (22/126*53=9.25)

I think changing the implementation of oracle will cause more confusion that what is solves, but the other implementation docs can be likely be adjusted to better reflect the implementation.

#9704 is likely worth implementing. Given that  currently the float used by oracle needs to be manually specied using with variant, it wouldn't need a depreciation warning for the current behaviour.

@zzzeek do you remember why the docs for float imply decimal precision in some parts and binary precision in others?
